lib/utils/sortcache/sortcache.go

```diff
@@ -0,0 +1,234 @@
+// be modified, and it is generally best to never modify stored resources.
+type SortCache[T any] struct {
+	rw      sync.RWMutex
+	indexes map[string]func(T) string
+	trees   map[string]*btree.BTreeG[entry]
+	values  map[uint64]T
+	counter uint64
+}
+
+type entry struct {
+	key string
+	ref uint64
+}
+
+// New sets up a new [SortCache] based on the provided configuration.
+func New[T any](cfg Config[T]) *SortCache[T] {
+	const (
+		// bTreeDegree of 8 is standard across most of the teleport codebase
+		bTreeDegree = 8
+	)
+
+	trees := make(map[string]*btree.BTreeG[entry], len(cfg.Indexes))
+
+	for index := range cfg.Indexes {
+		trees[index] = btree.NewG(bTreeDegree, func(a, b entry) bool {
+			return a.key < b.key
+		})
+	}
+
+	return &SortCache[T]{
+		indexes: cfg.Indexes,
+		trees:   trees,
+		values:  make(map[uint64]T),
+	}
+}
+
+// Get loads the value associated with the given index and key. ok will be false if either
+// the index does not exist or no value maps to the provided key on that index. Note that
+// mutating a value such that any of its index keys change is not permissible and will result
+// in permanently bad state. To avoid this, any implementation that might mutate returned
+// values must clone them.
+func (c *SortCache[T]) Get(index, key string) (value T, ok bool) {
+	c.rw.RLock()
+	defer c.rw.RUnlock()
+
+	ref, ok := c.lookup(index, key)
+	if !ok {
+		return value, false
+	}
+
+	return c.values[ref], true
+}
+
+// HasIndex checks if the specified index is present in this cache.
+func (c *SortCache[T]) HasIndex(name string) bool {
+	// index map is treated as immutable once created, so no lock is required.
+	_, ok := c.indexes[name]
+	return ok
+}
+
+// KeyOf gets the key of the supplied value on the given index.
+func (c *SortCache[T]) KeyOf(index string, value T) string {
+	// index map is treated as immutable once created, so no lock is required.
+	fn, ok := c.indexes[index]
+	if !ok {
+		return ""
+	}
+	return fn(value)
+}
+
+// lookup is an internal helper that finds the unique reference id for a value given a specific
+// index/key. Must be called either under read or write lock.
+func (c *SortCache[T]) lookup(index, key string) (ref uint64, ok bool) {
+	tree, exists := c.trees[index]
+	if !exists {
+		return 0, false
+	}
+
+	entry, exists := tree.Get(entry{key: key})
+	if !exists {
+		return 0, false
+	}
+
+	return entry.ref, true
+}
+
+// Put inserts a value into the sort cache, removing any existing values that collide with it. Since all indexes
+// are required to be unique, a single Put can end up evicting up to N existing values where N is the number of
+// indexes if it happens to collide with a different value on each index. implementations that expect evictions
+// to always either be zero or one (e.g. caches of resources with unique IDs) should check the returned eviction
+// count to help detect bugs.
+func (c *SortCache[T]) Put(value T) (evicted int) {
+	c.rw.Lock()
+	defer c.rw.Unlock()
+
+	c.counter++
```

Comment:
I like using a counter here instead of a key we assume would be unique (like in the unified resource cache). This counter shouldn't be assumed to be some unbroken sequential ID and just used in creating the keys, yes?

Reply:
If I understand you correctly, you're basically saying that we could expose the counter to the caller as a tool for making indexes where there were _never_ collisions.  E.g. instead of needing to add the resource name as a suffix for the `Kind` index, you could just include the counter, making it work like a non-unique column in a database?

I think thats a legitimate usecase, and the pattern we currently use, in the unified resource cache and in this PR, of always appending a unique suffix is a bit fiddly.  I like the idea of keeping the counter as an internal implementation detail though.

I think I'd prefer to actually explicitly add a new category of key rather than require the caller to fiddle with the IDs.  Taking the example of resource kinds mentioned above, we could do something like this:

```golang
cache := New(Config[Resource]{
    Indexes: map[string]func(Resource) string {
        "ResourceID": func(r Resource) string {
            return fmt.Sprintf("%s/%s", r.GetName(), r.GetKind())
        },
    },
    Attributes: map[string]func(Resource) string {
        "Kind": func(r Resource) string {
            return r.GetKind()
        },
    },
})

// ...

for _, node := range cache.GetAllWithAttr("Kind", "node") {
    doStuff(server)
}
```

Functionally equivalent, but it feels a bit cleaner to do it with a very explicit API.

Reply:
What I meant to say was "i like what you did here instead of how its done elsewhere because it doesn't cause collisions". I was just making sure that the counter doesn't need to be modified to be sequential, which I'm pretty sure it doesn't need to be sequential anyway (just making sure)